### PR TITLE
Frame stack

### PR DESCRIPTION
**Pull request by jackyoung96**
frame_stack.py modifying (same as stablebaseline3)
Original: fill observation stack by zero observations 
Modify: fill observation stack by the copy of the first observation

aec_mock_test modifying (Reflecting modified frame stack)

### CODE REVIEW

**Comment by benblack769** [comment]
@jackyoung96 Is this meant as a replacement for #141 ? If so, can you close that PR?

**Comment by benblack769** [comment]
This PR looks pretty good. I understand what it is supposed to do, and the changes are simple and minimal. 

Just a couple of additional things to make this production-ready:

* Could you fix the documentation here to reflect the new changes? https://github.com/Farama-Foundation/SuperSuit/blob/master/README.md?plain=1#L47
* Also, can you bump `frame_stack_v1` to `frame_stack_v2` (a simple search and replace throughout the repository should do the trick).

**Comment by jackyoung96** [comment]
> @jackyoung96 Is this meant as a replacement for #141 ? If so, can you close that PR?

Yes it is. Please close it.



> This PR looks pretty good. I understand what it is supposed to do, and the changes are simple and minimal.
> 
> Just a couple of additional things to make this production-ready:
> 
> * Could you fix the documentation here to reflect the new changes? https://github.com/Farama-Foundation/SuperSuit/blob/master/README.md?plain=1#L47
> * Also, can you bump `frame_stack_v1` to `frame_stack_v2` (a simple search and replace throughout the repository should do the trick).

I will. Thanks!

**Comment by jackyoung96** [comment]
Could you check it? Thank you!

**Comment by jackyoung96** [comment]
I add the stack_dim0 flag for stack_frame wrapper.

Pytorch uses channel first order and Tensorflow uses channel last order, so we can choose where the stacked dimension locate by stack_dim0 flag.

**Comment by jjshoots** [comment]
Hi @jackyoung96, can you fix linting and clear the test fails? Thanks!

**Comment by jackyoung96** [comment]
> Hi @jackyoung96, can you fix linting and clear the test fails? Thanks!

Sure, I will ASAP

**Comment by jackyoung96** [comment]
@benblack769 Sorry for the late commit for the lint test failure. I totally forgot about it. I finally push the final version!

**Comment by jjshoots** [comment]
@jackyoung96 Thanks, we will likely merge this once the Atari work is done :)

**Comment by jackyoung96** [comment]
This commit should be worked!

**Comment by jjshoots** [comment]
Hey @benblack769, sorry to drag you in, but do you know anything about the above error? I can't quite understand what's the error either.

**Comment by benblack769** [comment]
I belive it is just a flaky error that comes up from time to time. I think perhaps legal moves are just chosen by chance in Mahjog so the game doesn't end quickly enough for the test to pass in 1/1000 games or something. A bit silly.

**Comment by benblack769** [comment]
Just try rerunning the test

**Comment by jjshoots** [comment]
Hi @jackyoung96, tests are still failing, do you think that can be fixed?

**Comment by jjshoots** [comment]
@BolunDai0216

**Comment by BolunDai0216** [comment]
> @BolunDai0216

I think the tests are failing partially because of this line: https://github.com/Farama-Foundation/SuperSuit/blob/dca236d1b7924358c8fc4e14ca8b62f7c3739c8e/supersuit/generic_wrappers/frame_stack.py#L66

When creating `reset()` seed now needs to be added as an argument, i.e., `def reset(seed=None)`. @jackyoung96 Would it be possible to change it so that the `reset` matches `frame_stack_v1` argument-wise and see if that's the issue?

**Comment by jackyoung96** [comment]
@BolunDai0216 I fixed it and commit. I hope it can solve the issue.

File: supersuit/utils/frame_stack.py [hunk 1]
```diff
@@ -2,26 +2,41 @@
...
+            assert False, "Stacking is only avaliable for 1,2 or 3 dimentional arrays"
 
     return tile_shape, new_shape
 
 
-def stack_obs_space(obs_space, stack_size):
+def stack_obs_space(obs_space, stack_size, stack_dim0=False):
     """
     obs_space_dict: Dictionary of observations spaces of agents
     stack_size: Number of frames in the observation stack
```
Comment (jjshoots) [inline supersuit/utils/frame_stack.py:42]: Need a comment here describing what `stack_dim0` (or `stack_dim` possibly) does.

File: README.md [hunk 2]
```diff
@@ -44,7 +44,9 @@ You can install SuperSuit via `pip install supersuit`
...
+`frame_stack_v1(env, num_frames=4, stack_dim0=False)` stacks the most recent frames. For vector games observed via plain vectors (1D arrays), the output is just concatenated to a longer 1D array. 2D or 3D arrays are stacked to be taller 3D arrays. Stacked dimension can be set to dim=0 (default dim=-1) by stack_dim0=True. At the start of the game, frames that don't yet exist are filled with 0s. `num_frames=1` is analogous to not using this function.
+
+`frame_stack_v2(env, num_frames=4, stack_dim0=False)` stacks the most recent frames. For vector games observed via plain vectors (1D arrays), the output is just concatenated to a longer 1D array. 2D or 3D arrays are stacked to be taller 3D arrays. Stacked dimension can be set to dim=0 (default dim=-1) with `stack_dim0=True`. At the start of the game, frames that don't yet exist are filled with the copies of the first frame. `num_frames=1` is analogous to not using this function.
```
Comment (jjshoots) [inline README.md:49]: While I'm sure that it works, can we change `stack_dim0` to just become `stack_dim`, and then assert it to be either 0 or -1? This way if there are new API changes in the future that allow for arbitrary stacking dimensions, the arguments between the new API and this one won't be any different.

Also, do keep arguments (like `stack_dim0=True`) inside code lines.

File: supersuit/utils/agent_indicator.py [hunk 3]
```diff
@@ -74,13 +72,13 @@ def change_observation(obs, space, indicator_data):
...
             new_obs = np.pad(obs, (0, num_indicators))
-            new_obs[indicator_num + old_len] = 1.0
+            new_obs[indicator_num + old_len] = np.max(space.high)
             return new_obs
         elif ndims == 3 or ndims == 2:
             obs = obs if ndims == 3 else np.expand_dims(obs, 2)
             old_shaped3 = obs.shape[2]
             new_obs = np.pad(obs, [(0, 0), (0, 0), (0, num_indicators)])
-            new_obs[:, :, old_shaped3 + indicator_num] = 1.0
+            new_obs[:, :, old_shaped3 + indicator_num] = np.max(space.high)
```
Comment (jjshoots) [inline supersuit/utils/agent_indicator.py:81]: What's the reason for this change as well?

File: supersuit/generic_wrappers/frame_stack.py [hunk 4]
```diff
@@ -23,19 +23,70 @@ def modify_obs_space(self, obs_space):
...
+            if self.reset_flag:
+                for _ in range(stack_size):
+                    self.stack = stack_obs(
+                        self.stack,
+                        obs,
+                        self.old_obs_space,
+                        stack_size,
+                        stack_dim0
+                    )
+                self.reset_flag = False
```
Comment (jjshoots) [inline supersuit/generic_wrappers/frame_stack.py:80]: Is a for loop necessary here? Is it not possible to simply do something like `tile_shape * obs`? This is just to make things run a little bit faster.

I'm ok if a completely new function was created just for this functionality as well.

**Review by jjshoots — COMMENTED** [review]
Overall looks pretty good to me, just some general comments about code cleanliness some minor improvement. Once the comments are addressed, I'll get someone else to review it before merging.

Thanks for your contribution!

File: supersuit/utils/agent_indicator.py [hunk 5]
```diff
@@ -8,17 +8,15 @@ def change_obs_space(space, num_indicators):
     if isinstance(space, Box):
         ndims = len(space.shape)
         if ndims == 1:
-            pad_space = np.ones((num_indicators,), dtype=space.dtype)
+            pad_space = np.max(space.high) * np.ones((num_indicators,), dtype=space.dtype)
```
[the remark this answers is not in the document]
Comment (jackyoung96) [reply supersuit/utils/agent_indicator.py:11]: It unify the high and low range of the indicator channel with the original observation channels.
Because sometimes the values original observation channel are in range of [0,255] (Before normalized)

File: supersuit/utils/agent_indicator.py (hunk 3 above)
Comment (jackyoung96) [reply supersuit/utils/agent_indicator.py:81]: Same reason with https://github.com/Farama-Foundation/SuperSuit/pull/142#discussion_r865863503

File: supersuit/generic_wrappers/frame_stack.py (hunk 4 above)
Comment (jackyoung96) [reply supersuit/generic_wrappers/frame_stack.py:80]: Your right. Actually, this is how frame_stack_v0 was implemented before, so I left it as it is. If you want, we can modify it, but I think it makes the code (very x 3) little faster 😅

File: supersuit/utils/agent_indicator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjshoots) [reply supersuit/utils/agent_indicator.py:11]: Is it possible to instead do `np.min(space.high)`? This is in the very rare but potentially possible case of `space.high = [1, 255, 255]`, `pad_space` will then not be contained within the observation space.

You could implement it to check that all `space.high` are the same values `np.all(space.high[0] == space.high)`, and if they aren't then raise a warning. Doing this within the wrapper function would incur overhead cost though, so I think it's better to put this assert in the wrapper init, and then just do `np.min(space.high)` everywhere else.

File: supersuit/generic_wrappers/frame_stack.py (hunk 4 above)
Comment (jjshoots) [reply supersuit/generic_wrappers/frame_stack.py:80]: If it makes the code a lot less readable then we can just keep the original implementation, but if it's simple to implement, I think changing it would be better. Small optimizations like this add up a lot in the long run (especially for python). :smile:

**Comment by jjshoots** [comment]
Hi @jackyoung96, tests are still failing, though I think it's just linting.
